core/embed/rust/src/smp/image_info.rs

```diff
@@ -0,0 +1,202 @@
+use super::{
+    receiver_acquire, receiver_release, send_request, wait_for_response, MsgType, SmpBuffer,
+    SmpHeader, SMP_CMD_ID_IMAGE_STATE, SMP_GROUP_IMAGE, SMP_HEADER_SIZE, SMP_OP_READ,
+};
+use crate::time::Duration;
+use minicbor::{data::Type, decode, Decoder, Encoder};
+
+/// MCUboot-compatible version structure matching image header format
+#[derive(Clone, Copy, Debug)]
+pub struct AppVersion {
+    pub major: u8,
+    pub minor: u8,
+    pub revision: u16,
+    pub build_num: u32,
+}
+
+/// Macro to generate overflow-safe decimal parsers for different integer types
+macro_rules! make_parser {
+    ($name:ident, $target:ty, $accumulator:ty) => {
+        fn $name(seg: &str) -> Option<$target> {
+            if seg.is_empty() {
+                return None;
+            }
+            let mut val: $accumulator = 0;
+            for b in seg.as_bytes() {
+                if *b < b'0' || *b > b'9' {
+                    return None;
+                }
+                val = val
+                    .checked_mul(10)?
+                    .checked_add((b - b'0') as $accumulator)?;
+                if val > <$target>::MAX as $accumulator {
+                    return None;
+                }
+            }
+            Some(val as $target)
+        }
+    };
+}
+
+// Generate the three parser functions
+make_parser!(parse_u8, u8, u16);
+make_parser!(parse_u16, u16, u32);
+make_parser!(parse_u32, u32, u64);
+
+/// Parse "<major>.<minor>.<revision>[.<build>]" into AppVersion (no allocation)
+/// Matches MCUboot image header version format
+/// Examples: "1.0.3", "1.0.3.0", "255.255.65535.4294967295"
+fn parse_app_version(s: &str) -> Option<AppVersion> {
+    let mut parts = s.split('.');
+
+    let major = parse_u8(parts.next()?)?;
+    let minor = parse_u8(parts.next()?)?;
+    let revision = parse_u16(parts.next()?)?;
+
+    // Build number is optional (defaults to 0 if absent)
+    let build_num = match parts.next() {
```

Comment:
somewhat more rustier version of writing this could be:     let build_num = parts
.next()
.filter(|b| !b.is_empty())
.map_or(Some(0), parse_u32)?;